CliWrap/PipeSource.cs

```diff
@@ -130,4 +152,79 @@ await _command
             .WithStandardOutputPipe(PipeTarget.ToStream(destination))
             .ExecuteAsync(cancellationToken)
             .ConfigureAwait(false);
+}
+
+internal class StreamDelegatePipeSource : PipeSource
+{
+    private readonly Action<Stream> _source;
+
+    public StreamDelegatePipeSource(Action<Stream> source) => _source = source;
+
+    public override Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+    {
+        _source(destination);
+        return Task.CompletedTask;
+    }
+}
+
+internal class StreamAsyncDelegatePipeSource : PipeSource
+{
+    private readonly Func<Stream, CancellationToken, Task> _source;
+
+    public StreamAsyncDelegatePipeSource(Func<Stream, CancellationToken, Task> source) => _source = source;
+
+    public override async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default) =>
+        await _source(destination, cancellationToken).ConfigureAwait(false);
+}
+
+internal class TextWriterDelegatePipeSource : PipeSource
```

Comment:
Can we avoid adding `TextWriterDeleaget...`? I think it's already trivial to create a text writer using the other overload (as you've shown in tests) and it allows the user to pass the configuration (such as bufferSize, etc.) directly instead of routing it through us.